src/bson_mapper/bson_archiver.hpp

```diff
@@ -0,0 +1,1078 @@
+        : OutputArchive<BSONOutputArchive>{this},
+          _bsonBuilder{false},
+          _writeStream{stream},
+          _nextName{nullptr} {
+    }
+
+   private:
+    /**
+     * Writes the current contents of the BSON document builder to the
+     * output stream.
+     */
+    void writeDoc() {
+        _writeStream.write(reinterpret_cast<const char*>(_bsonBuilder.view_document().data()),
+                           _bsonBuilder.view_document().length());
+        _bsonBuilder.clear();
+    }
+
+   public:
+    /**
+     * Checks if the most recent object written was in the root of the document.
+     * If so, that document is written to the archive.
+     */
+    void writeDocIfRoot() {
+        if (_nodeTypeStack.empty()) {
+            writeDoc();
+        }
+    }
+
+    /**
+     * Starts a new node in the BSON output.
+     *
+     * The node can optionally be given a name by calling setNextName prior to
+     * creating the node. Nodes only need to be started for types that are
+     * objects or arrays.
+     */
+    void startNode() {
+        writeName(true);
+        _nodeTypeStack.push(OutputNodeType::StartObject);
+    }
+
+    /**
+     * Designates the most recently added node as finished.
+     */
+    void finishNode() {
+        // Iff we ended up serializing an empty object or array, writeName
+        // will never have been called - so start and then immediately end
+        // the object/array.
+        //
+        // We'll also end any object/arrays we happen to be in.
+
+        bool closedObj = false;
+        switch (_nodeTypeStack.top()) {
+            case OutputNodeType::StartArray:
+                _bsonBuilder.open_array();
+            case OutputNodeType::InArray:
+                _bsonBuilder.close_array();
+                closedObj = true;
+                break;
+            case OutputNodeType::StartObject:
+                if (_nodeTypeStack.size() > 1) {
+                    _bsonBuilder.open_document();
+                }
+            case OutputNodeType::InObject:
+                if (_nodeTypeStack.size() > 1) {
+                    _bsonBuilder.close_document();
+                }
+                closedObj = true;
+                break;
+        }
+
+        _nodeTypeStack.pop();
+
+        if (closedObj) {
+            writeDocIfRoot();
+        }
+    }
+
+    /**
+     * Sets the name for the next node or element.
+     *
+     * @param name
+     *    The name for the next node or element.
+     */
+    void setNextName(const char* name) {
+        _nextName = name;
+    }
+
+    /**
+     * Implementations of saveValue that save a BSON type to the current node,
+     * for every BSON type except those that are deprecated and those that are
+     * reserved for internal use.
+     *
+     * @param val
+     *    The BSON typed value to save to the archive.
+     */
+    template <class BsonT, traits::EnableIf<is_bson<BsonT>::value> = traits::sfinae>
+    void saveValue(BsonT val) {
```

Comment:
I wonder if you don't want to just have the single overload with anything the core builder will take.

It already has a variety of overloads for append that cover all the overloads that are possible